tests/RetryMiddlewareTest.php

```diff
@@ -71,10 +72,36 @@ public function testCanRetryExceptions()
 
     public function testBackoffCalculateDelay()
     {
-        self::assertSame(0, RetryMiddleware::exponentialDelay(0));
-        self::assertSame(1000, RetryMiddleware::exponentialDelay(1));
-        self::assertSame(2000, RetryMiddleware::exponentialDelay(2));
-        self::assertSame(4000, RetryMiddleware::exponentialDelay(3));
-        self::assertSame(8000, RetryMiddleware::exponentialDelay(4));
+        $calls = [];
+        $decider = function ($retries, $request, $response, $error) use (&$calls) {
+            $calls[] = \func_get_args();
+            return $error instanceof \Exception;
+        };
+        $m = Middleware::retry($decider);
+        $h = new MockHandler();
+        $c = new Client(['handler' => $m($h)]);
+
+        $h->append(new \Exception(), new Response());
+        $p = $c->send(new Request('GET', 'http://test.com'), []);
+
+        self::assertSame(200, $p->getStatusCode());
+        self::assertCount(2, $calls);
+        self::assertSame(1000, $h->getLastOptions()['delay']);
+
+        $calls = [];
+        $h->append(new \Exception(), new \Exception(), new Response());
+        $p = $c->send(new Request('GET', 'http://test.com'), []);
+
+        self::assertSame(200, $p->getStatusCode());
+        self::assertCount(3, $calls);
+        self::assertSame(2000, $h->getLastOptions()['delay']);
+
+        $calls = [];
+        $h->append(new \Exception(), new \Exception(), new \Exception(), new Response());
+        $p = $c->send(new Request('GET', 'http://test.com'), []);
+
+        self::assertSame(200, $p->getStatusCode());
+        self::assertCount(4, $calls);
+        self::assertSame(4000, $h->getLastOptions()['delay']);
```

Comment:
My main problem here is that these tests are causing actual delays.. any workaround would be helpful here.

Reply:
You could override the sleep function, or create an injectable/overrideable clock that falls back to the system clock. But I think this is fine for now as it is.

Reply:
Or.. we could just keep it public.

The tests were easier to read and write and did not take forever to run =)

Reply:
Or...we could not test it. 😄 It's a trivial sleep after all...

Reply:
There has recently been a bug in this trivial code. I think the tests should be there.

Reply:
Let's not have this delay v7. Will add it to v8 milestone...